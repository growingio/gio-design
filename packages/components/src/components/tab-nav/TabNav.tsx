import React, { useMemo, useState, useEffect, useRef } from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import { TabNavProps, TabNavItemProps } from './interface';
import useRefs from '../../utils/hooks/useRefs';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

const TabNav = (props: TabNavProps, ref?: React.RefObject<HTMLDivElement>) => {
  const {
    children,
    prefixCls: customizePrefixCls,
    type = 'block',
    size = 'large',
    onChange,
    onTabClick,
    activeKey,
    defaultActiveKey = '',
  } = props;

  const [localActiveKey, setLocalActiveKey] = useState<string | number>(activeKey || defaultActiveKey);
  const [inkStyle, setInkStyle] = useState<{ left?: number; width?: number }>({});
  const wrapperRefKey = useRef<symbol>(Symbol('tabNav'));
  const [setRef, getRef] = useRefs<HTMLDivElement>();

  const prefixCls = usePrefixCls('tabnav', customizePrefixCls);
  const classString = classNames(prefixCls, `${prefixCls}-${type}`, {
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-xs`]: size === 'xs' && type === 'block',
  });

  const [tabNavKeys, tabNavChildren] = useMemo(() => {
    const _tabNavKeys: (string | number)[] = [];
    const _tabNavChildren = toArray(children).map((node: React.ReactElement<TabNavItemProps>, index) => {
      if (React.isValidElement(node) && node.type === TabNav.Item) {
        const { className, disabled, onClick, ...rest } = node.props;
        const _key = isNil(node.key) ? index : node.key;
        _tabNavKeys.push(_key);
        return (
          <TabNav.Item
            className={classNames(className, `${prefixCls}-item`, {
              [`${prefixCls}-item-active`]: localActiveKey === _key,
              [`${prefixCls}-item-disabled`]: disabled,
            })}
            prefixCls={prefixCls}
            disabled={disabled}
            key={_key}
            ref={setRef(_key)}
            onClick={(e) => {
              if (!disabled) {
                onClick?.(e);
                onTabClick?.(_key);
                if (isNil(activeKey)) {
                  setLocalActiveKey(_key);
                }
                if (localActiveKey !== _key) {
                  onChange?.(_key);
                }
              }
            }}
            {...rest}
          />
        );
      }
      return null;
    });
    return [_tabNavKeys, _tabNavChildren];
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children, localActiveKey, onChange, onTabClick]);

  useMemo(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, [localActiveKey, tabNavKeys]);

  useMemo(() => {
    if (!isNil(activeKey) && tabNavKeys.includes(activeKey)) {
      setLocalActiveKey(activeKey);
    }
  }, [activeKey, tabNavKeys]);

  useEffect(() => {
    if (!isNil(getRef(localActiveKey)?.current) && !isNil(getRef(wrapperRefKey.current)?.current)) {
      const { left, width } = getRef(localActiveKey)!.current!.getBoundingClientRect();
      const wrapperLeft = getRef(wrapperRefKey.current)!.current!.getBoundingClientRect().left;
      setInkStyle({ left: left - wrapperLeft, width });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [localActiveKey, tabNavChildren]);

  return (
    <div className={classString} ref={setRef(wrapperRefKey.current, ref)}>
      {tabNavChildren}
      <div className={`${prefixCls}-ink-bar`} style={{ ...inkStyle }} />
    </div>
  );
};

TabNav.Item = React.forwardRef(
  ({ prefixCls, children, ...rest }: TabNavItemProps, ref: React.RefObject<HTMLDivElement>) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <div ref={ref} {...rest}>
      <div className={`${prefixCls}-item-btn`}>{children}</div>
    </div>
  )
);

export default TabNav;
