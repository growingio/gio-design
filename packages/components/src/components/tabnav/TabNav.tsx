import React, {
  useContext, useMemo, useState, useEffect, useRef,
} from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import { ConfigContext } from '../config-provider';
import { TabNavProps, TabNavItemProps } from './interface';
import useRefs from '../../utils/hooks/useRefs';

const TabNav = (props: TabNavProps, ref?: React.RefObject<HTMLDivElement>) => {
  const {
    children,
    prefixCls: customizePrefixCls,
    type = 'block',
    size = 'large',
    onChange,
    onTabClick,
    activeKey,
    defaultActiveKey = '',
  } = props;

  const [localActiveKey, setLocalActiveKey] = useState<string | number>(activeKey || defaultActiveKey);
  const [inkStyle, setInkStyle] = useState<{ left?: number; width?: number }>({});
  const wrapperRefKey = useRef<symbol>(Symbol('tabNav'));
  const [setRef, getRef] = useRefs<HTMLDivElement>();

  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('tabnav', customizePrefixCls);
  const classString = classNames(prefixCls, `${prefixCls}-${type}`, {
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-xs`]: size === 'xs' && type === 'block',
  });

  const [tabNavKeys, tabNavChildren] = useMemo(() => {
    const _tabNavKeys: (string | number)[] = [];
    const _tabNavChildren = toArray(children).map((node: React.ReactElement<TabNavItemProps>, index) => {
      if (React.isValidElement(node) && node.type === TabNav.Item) {
        const {
          className, disabled, onClick, ...rest
        } = node.props;
        const _key = isNil(node.key) ? index : node.key;
        _tabNavKeys.push(_key);
        return (
          <TabNav.Item
            className={classNames(className, `${prefixCls}-item`, {
              [`${prefixCls}-item-active`]: localActiveKey === _key,
              [`${prefixCls}-item-disabled`]: disabled,
            })}
            prefixCls={prefixCls}
            disabled={disabled}
            key={_key}
            ref={setRef(_key)}
            onClick={(e) => {
              if (!disabled) {
                onClick?.(e);
                onTabClick?.(_key);
                if (isNil(activeKey)) {
                  setLocalActiveKey(_key);
                }
                if (localActiveKey !== _key) {
                  onChange?.(_key);
                }
              }
            }}
            {...rest}
          />
        );
      }
      return null;
    });
    return [_tabNavKeys, _tabNavChildren];
  }, [children, localActiveKey, onChange, onTabClick]);

  useMemo(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, []);

  useMemo(() => {
    if (!isNil(activeKey) && tabNavKeys.includes(activeKey)) {
      setLocalActiveKey(activeKey);
    }
  }, [activeKey]);

  useEffect(() => {
    if (!isNil(getRef(localActiveKey)?.current) && !isNil(getRef(wrapperRefKey.current)?.current)) {
      const { left, width } = getRef(localActiveKey)!.current!.getBoundingClientRect();
      const wrapperLeft = getRef(wrapperRefKey.current)!.current!.getBoundingClientRect().left;
      setInkStyle({ left: left - wrapperLeft, width });
    }
  }, [localActiveKey, tabNavChildren]);

  return (
    <div className={classString} ref={setRef(wrapperRefKey.current, ref)}>
      {tabNavChildren}
      <div className={`${prefixCls}-ink-bar`} style={{ ...inkStyle }} />
    </div>
  );
};

TabNav.Item = React.forwardRef(
  ({ prefixCls, children, ...rest }: TabNavItemProps, ref: React.RefObject<HTMLDivElement>) => (
    <div ref={ref} {...rest}>
      <div className={`${prefixCls}-item-btn`}>{children}</div>
    </div>
  ),
);

export default TabNav;
