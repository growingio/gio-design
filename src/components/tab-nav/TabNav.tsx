/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { useRef, useMemo, useEffect } from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import { TabNavProps, TabNavItemProps } from './interface';
import useRefs from '../../utils/hooks/useRefs';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import useControlledState from '../../utils/hooks/useControlledState';

const TabNav = (props: TabNavProps, ref?: React.RefObject<HTMLDivElement>) => {
  const {
    children,
    prefixCls: customizePrefixCls,
    type = 'block',
    size = 'large',
    onChange,
    onTabClick,
    activeKey,
    defaultActiveKey = '',
  } = props;

  const [localActiveKey, setLocalActiveKey] = useControlledState<string>(activeKey, defaultActiveKey);
  const wrapperRefKey = useRef<symbol>(Symbol('tabNav'));
  const [setRef, getRef] = useRefs<HTMLDivElement>();

  const prefixCls = usePrefixCls('tabnav', customizePrefixCls);
  const classString = classNames(prefixCls, `${prefixCls}-${type}`, {
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-xs`]: size === 'xs' && type === 'block',
  });

  const [tabNavKeys, tabNavChildren] = useMemo(() => {
    const _tabNavKeys: string[] = [];
    const _tabNavChildren =
      toArray(children)
      .filter((node) => React.isValidElement(node) && node.type === TabNav.Item)
      .map((node, index) => {
        const { className, disabled, onClick, ...rest } = node.props;
        const _key: string = isNil(node.key) ? index.toString() : node.key.toString();
        _tabNavKeys.push(_key);
        return React.cloneElement(node, {
          className: classNames(className, `${prefixCls}-item`, {
            [`${prefixCls}-item-active`]: localActiveKey === _key,
            [`${prefixCls}-item-disabled`]: disabled,
          }),
          prefixCls,
          disabled,
          key: _key,
          ref: setRef(_key),
          onClick: (e: React.MouseEvent<HTMLDivElement>) => {
            if (!disabled) {
              onClick?.(e);
              onTabClick?.(_key);
              setLocalActiveKey(_key);
              if (localActiveKey !== _key) {
                onChange?.(_key);
              }
            }
          },
          ...rest,
        });
      });
    return [_tabNavKeys, _tabNavChildren];
  }, [children, localActiveKey, onChange, onTabClick, prefixCls, setLocalActiveKey, setRef]);

  useEffect(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, [tabNavKeys, localActiveKey, setLocalActiveKey]);

  const inkStyle = useMemo(() => {
    if (isNil(getRef(localActiveKey)?.current) || isNil(getRef(wrapperRefKey?.current)?.current)) {
      return {};
    }
    const { left, width } = getRef(localActiveKey)!.current!.getBoundingClientRect();
    const wrapperLeft = getRef(wrapperRefKey.current)!.current!.getBoundingClientRect().left;
    return { left: left - wrapperLeft, width };
  } , [getRef, localActiveKey]);

  return (
    <div className={classString} ref={setRef(wrapperRefKey.current, ref)}>
      {tabNavChildren}
      <div className={`${prefixCls}-ink-bar`} style={{ ...inkStyle }} />
    </div>
  );
};

TabNav.Item = React.forwardRef<HTMLDivElement, TabNavItemProps>(({ prefixCls, children, ...rest }, ref) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <div data-testid='tabnav-item' ref={ref} {...rest}>
    <div className={`${prefixCls}-item-btn`}>{children}</div>
  </div>
));

export default TabNav;
