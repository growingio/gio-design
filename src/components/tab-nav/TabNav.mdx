import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import TabNav from './TabNav';

# TabNav 标签导航

<Subtitle>标签导航是两个或更多按钮段的集合。标签导航 TabNav 与 标签页 Tabs 的区别：点击切换标签页时，页面内容随之切换；点击分段控件，切换内容由页面决定。</Subtitle>

## 参数说明
<ArgsTable of={TabNav} />

## 代码演示

### 块状标签导航
一共四种尺寸：L（40px）、M（36px）、S（30px）、XS（24px）。

[Example](/?path=/story/basic-components-avatar--block)

### 线状标签导航
一共三种尺寸：L（40px）、M（36px）、S（30px）。

[Example](/?path=/story/basic-components-avatar--line)
