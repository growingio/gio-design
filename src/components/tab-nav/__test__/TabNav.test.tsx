import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TabNav from '../index';

describe('Testing TabNav', () => {
  const getTabNav = (args?: unknown) => (
    <TabNav {...args}>
      <TabNav.Item key='1'>111</TabNav.Item>
      <TabNav.Item key='2'>222</TabNav.Item>
      <TabNav.Item key='3'>333</TabNav.Item>
      <TabNav.Item key='4' disabled>444</TabNav.Item>
    </TabNav>
  );
    
  it('should be stable', () => {
    const { asFragment } = render(getTabNav());
    expect(asFragment()).toMatchSnapshot();
  });

  it('will be changed after click', () => {
    const { getByText, container } = render(getTabNav());
    fireEvent.click(getByText('222'));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('222');
    fireEvent.click(getByText('444'));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).not.toBe('444');
    expect(container.getElementsByClassName('gio-tabnav-item-disabled')[0].textContent).toBe('444');
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const { rerender, unmount } = render(getTabNav());
      rerender(getTabNav({ type: 'line' }));
      rerender(getTabNav({ size: 'small' }));
      unmount()
    }).not.toThrow();
  });

  test('props type', () => {
    const { container, rerender } = render(getTabNav());
    expect(container.getElementsByClassName('gio-tabnav-block')).not.toBe(null);
    rerender(getTabNav({ type: 'line' }));
    expect(container.getElementsByClassName('gio-tabnav-line')).not.toBe(null);
  });

  test('props size', () => {
    const { container, rerender } = render(getTabNav());
    expect(container.getElementsByClassName('gio-tabnav-lg')).not.toBe(null);
    rerender(getTabNav({ size: 'middle' }));
    expect(container.getElementsByClassName('gio-tabnav-md')).not.toBe(null);
    rerender(getTabNav({ size: 'small' }));
    expect(container.getElementsByClassName('gio-tabnav-sm')).not.toBe(null);
  });

  test('prop defaultActiveKey', () => {
    const { container } = render(getTabNav({ defaultActiveKey: '2'}));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('222');
  });

  test('prop activeKey', () => {
    const { container, getByText, rerender } = render(getTabNav({ activeKey: '1' }));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('111');
    rerender(getTabNav({ activeKey: '2' }));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('222');
    // be controlled, not work
    fireEvent.click(getByText('111'))
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('222');
  });

  test('prop onChange and onTabClick', () => {
    const onChange = jest.fn();
    const onTabClick = jest.fn();
    const { getByText } = render(getTabNav({ onChange, onTabClick }));
    fireEvent.click(getByText('111'));
    // not change
    expect(onChange).not.toHaveBeenCalled();
    expect(onTabClick).toHaveBeenCalled();
    fireEvent.click(getByText('222'));
    expect(onChange).toHaveBeenCalled();
  });

  it('should be render rightly', () => {
    const { container } = render(getTabNav());
    expect(container.getElementsByClassName('gio-tabnav-item')).toHaveLength(4);
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('111');
    expect(container.getElementsByClassName('gio-tabnav-item-disabled')[0].textContent).toBe('444');
  });

  it('should be render content rightly', () => {
    const { container, getByText } = render(getTabNav());
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('111');
    fireEvent.click(getByText('222'));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('222');
  });

  it('only render TabNav.Item', () => {
    const { queryByText } = render(
      <TabNav defaultActiveKey="1">
        <TabNav.Item>111</TabNav.Item>
        <TabNav.Item>222</TabNav.Item>
        <TabNav.Item>333</TabNav.Item>
        <TabNav.Item disabled>444</TabNav.Item>
        <div>我不会被渲染</div>
      </TabNav>
    );
    expect(queryByText('我不会被渲染')).toBe(null);
  });
});
