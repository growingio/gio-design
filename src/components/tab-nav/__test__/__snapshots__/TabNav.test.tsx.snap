// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TabNav should be stable 1`] = `
<DocumentFragment>
  <div
    class="gio-tabnav gio-tabnav-block gio-tabnav-lg"
  >
    <div
      class="gio-tabnav-item gio-tabnav-item-active"
      data-testid="tabnav-item"
    >
      <div
        class="gio-tabnav-item-btn"
      >
        111
      </div>
    </div>
    <div
      class="gio-tabnav-item"
      data-testid="tabnav-item"
    >
      <div
        class="gio-tabnav-item-btn"
      >
        222
      </div>
    </div>
    <div
      class="gio-tabnav-item"
      data-testid="tabnav-item"
    >
      <div
        class="gio-tabnav-item-btn"
      >
        333
      </div>
    </div>
    <div
      class="gio-tabnav-item gio-tabnav-item-disabled"
      data-testid="tabnav-item"
      disabled=""
    >
      <div
        class="gio-tabnav-item-btn"
      >
        444
      </div>
    </div>
    <div
      class="gio-tabnav-ink-bar"
      style="left: 0px; width: 0px;"
    />
  </div>
</DocumentFragment>
`;
